mirrord-layer/src/pod_api.rs

```diff
@@ -412,8 +390,342 @@ async fn create_job_pod_agent(
+            .list(&ListParams::default().labels(&format!("app={}", pod_label)))
+            .await
+            .map_err(LayerError::KubeError)?;
+
+        let first_pod = deployment_pods.items.first().ok_or_else(|| {
+            LayerError::DeploymentNotFound(format!(
+                "Failed to fetch the default(first pod) from ObjectList<Pod> for {}",
+                self.deployment.clone()
+            ))
+        })?;
+
+        let pod_name = first_pod.clone().metadata.name.ok_or_else(|| {
+            LayerError::DeploymentNotFound(format!(
+                "Failed to fetch the name of the default pod in deployment {}, pod {:?}",
+                self.deployment.clone(),
+                first_pod.clone()
+            ))
+        })?;
+
+        let (
+            ContainerRuntime {
+                container_runtime,
+                socket_path,
+                container_id,
+            },
+            container_name,
+        ) = || -> Option<(String, String)> {
+            let container_runtime_and_id = first_pod
+                .clone()
+                .status?
+                .container_statuses?
+                .first()?
+                .container_id
+                .clone()?;
+            let container_name = first_pod
+                .clone()
+                .spec?
+                .containers
+                .first()
+                .map(|container| container.name.clone())?;
+
+            Some((container_runtime_and_id, container_name))
+        }()
+        .ok_or_else(|| {
+            LayerError::ContainerNotFound(format!(
+                "Failed to fetch container for pod {:?}, deployment {}",
+                first_pod.clone(),
+                self.deployment.clone()
+            ))
+        })
+        .and_then(|(runtime_and_id, container_name)| {
+            Ok((runtime_and_id.parse::<ContainerRuntime>()?, container_name))
+        })?;
+
+        let node_name =
+            || -> Option<String> { first_pod.clone().spec?.node_name }().ok_or_else(|| {
+                LayerError::NodeNotFound(format!(
+                    "target {:?} pod {:?}, container {}",
+                    self.clone(),
+                    first_pod.clone(),
+                    container_name.clone()
+                ))
+            })?;
+
+        Ok(RuntimeData {
+            node_name,
+            pod_name,
+            container_name,
+            container_id,
+            container_runtime,
+            socket_path,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum Target {
+    Deployment(DeploymentData),
+    Pod(PodData),
+}
+
+#[async_trait]
+trait RuntimeDataProvider {
+    async fn runtime_data(&self, client: &Client, namespace: &str) -> Result<RuntimeData>;
+}
+
+impl Target {
+    pub async fn runtime_info(&self, client: &Client, namespace: &str) -> Result<RuntimeData> {
+        match self {
+            Target::Pod(pod) => pod.runtime_data(client, namespace).await,
```

Comment:
Now that they both take the same params, can you hide the function behind a trait?